substrate/frame/revive/rpc/src/client/runtime_api.rs

```diff
@@ -65,9 +70,37 @@ impl RuntimeApi {
 	pub async fn dry_run(
 		&self,
 		tx: GenericTransaction,
+		block: BlockNumberOrTagOrHash,
 	) -> Result<EthTransactInfo<Balance>, ClientError> {
-		let payload = subxt_client::apis().revive_api().eth_transact(tx.into());
-		let result = self.0.call(payload).await?;
+		let timestamp_override = match block {
+			BlockNumberOrTagOrHash::BlockTag(BlockTag::Pending) =>
+				Some(Timestamp::current().as_millis()),
+			_ => None,
+		};
+
+		let payload = subxt_client::apis()
+			.revive_api()
+			.eth_transact_with_config(
+				tx.clone().into(),
+				DryRunConfig::new(timestamp_override).into(),
+			)
+			.unvalidated();
+
+		let result = self
+			.0
+			.call(payload)
+			.or_else(|err| async {
+				match err {
+					Metadata(MetadataError::RuntimeMethodNotFound(name)) => {
+						log::debug!(target: LOG_TARGET, "Method {name:?} not found falling back to eth_transact");
+						let payload = subxt_client::apis().revive_api().eth_transact(tx.into());
+						self.0.call(payload).await
+					},
+					e => Err(e),
```

Comment:
Does this also cover the historical block case where the latest node metadata has the new function? The `MetadataError::RuntimeMethodNotFound(name)`  suggests it does not. Maybe just use the new function for `pending` blocks for now if that helps.

Reply:
But this is bad? We want it to retry for a historical block? Not fail. So we should add another match arm for the error you just posted?